zkevm-circuits/src/copy_circuit.rs

```diff
@@ -324,26 +326,81 @@ impl<F: Field> SubCircuitConfig<F> for CopyCircuitConfig<F> {
                 },
             );
 
-            // addr change: for tx log, addr use addr_slot as index, not increase by 1
+            // for all cases, rw_counter + rwc_inc_left keeps same
             cb.condition(
                 and::expr([
                     not_last_two_rows.expr(),
-                    not::expr(tag.value_equals(CopyDataType::Padding, Rotation::cur())(
-                        meta,
-                    )),
-                    not::expr(tag.value_equals(CopyDataType::TxLog, Rotation::cur())(meta)),
+                    not::expr(tag.value_equals(CopyDataType::Padding, Rotation::cur())(meta)),
                 ]),
                 |cb| {
                     cb.require_equal(
-                        "rows[0].addr + 1 == rows[2].addr",
-                        meta.query_advice(addr, Rotation::cur()) + 1.expr(),
-                        meta.query_advice(addr, Rotation(2)),
+                        "rows[0].rw_counter + rows[0].rwc_inc_left == rows[1].rw_counter + rows[1].rwc_inc_left",
+                        meta.query_advice(rw_counter, Rotation::cur()) + meta.query_advice(rwc_inc_left, Rotation::cur()),
+                        meta.query_advice(rw_counter, Rotation::next()) + meta.query_advice(rwc_inc_left, Rotation::next()),
                     );
-                },
+                }
             );
+            // for all cases, rows[0].rw_counter + diff == rows[1].rw_counter
+            cb.condition(
+                and::expr([
+                    is_word_continue.is_lt(meta, None),
+                    not_last_two_rows.expr(),
+                    not::expr(tag.value_equals(CopyDataType::Padding, Rotation::cur())(meta)),
+                ]),
+                |cb| {
+                    let is_memory2memory = and::expr([
+                        meta.query_advice(is_memory, Rotation::cur()),
+                        meta.query_advice(is_memory, Rotation::next()),
+                    ]);
+                    let diff = select::expr(
+                        is_memory2memory,
+                        select::expr(meta.query_selector(q_step), 1.expr(), -1.expr()),
```

Comment:
What are the `1` and `-1` here?

Reply:
it's arranged as a read row, a write row, ...
for memory write it has its own rw counter by increase 1

Reply:
Still not clear, when can the counter go backwards (-1)? 

The most important requirement of the copy circuit is this:
- It looks up an RW operation for each of `[rwc .. rwc + rwc_left]`

Can this be made simpler or clearer how this is enforced?

Reply:
check this https://docs.google.com/spreadsheets/d/1UyNBdtFM1a6V_amGEc5UhNKLu1eYW2NgXbmp8v-PYvw/edit?usp=sharing

rw counter starts from 0x7c for the first read step and then a write step of 0x7d, so for the next read row, it's need to be same as the fist read row.

constraint using this way is for simplify the condition and reduce the degree.

Reply:
@naure if you have no more comments, I will merge this PR into memory_opt soon.